// @criteria/json-schema-validation/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
exports.__esModule = true;
exports.ValidationError = exports.validateJSONDraft2020_12 = exports.validateJSONAsyncDraft2020_12 = exports.jsonValidatorDraft2020_12 = exports.jsonValidatorAsyncDraft2020_12 = exports.isJSONValidDraft2020_12 = exports.isJSONValidAsyncDraft2020_12 = exports.validateJSONDraft07 = exports.validateJSONAsyncDraft07 = exports.jsonValidatorDraft07 = exports.jsonValidatorAsyncDraft07 = exports.isJSONValidDraft07 = exports.isJSONValidAsyncDraft07 = exports.validateJSONDraft06 = exports.validateJSONAsyncDraft06 = exports.jsonValidatorDraft06 = exports.jsonValidatorAsyncDraft06 = exports.isJSONValidDraft06 = exports.isJSONValidAsyncDraft06 = exports.validateJSONDraft04 = exports.validateJSONAsyncDraft04 = exports.jsonValidatorDraft04 = exports.jsonValidatorAsyncDraft04 = exports.isJSONValidDraft04 = exports.isJSONValidAsyncDraft04 = exports.validateJSONAsync = exports.validateJSON = exports.jsonValidatorAsync = exports.jsonValidator = exports.isJSONValidAsync = exports.isJSONValid = void 0;
var ValidationError_1 = require("./errors/ValidationError");
exports.ValidationError = ValidationError_1.ValidationError;
var jsonValidator_1 = require("./validation/jsonValidator");
exports.isJSONValid = jsonValidator_1.isJSONValid;
exports.isJSONValidAsync = jsonValidator_1.isJSONValidAsync;
exports.jsonValidator = jsonValidator_1.jsonValidator;
exports.jsonValidatorAsync = jsonValidator_1.jsonValidatorAsync;
exports.validateJSON = jsonValidator_1.validateJSON;
exports.validateJSONAsync = jsonValidator_1.validateJSONAsync;
var draft_04_1 = require("./specification/draft-04");
__createBinding(exports, draft_04_1, "isJSONValidAsync", "isJSONValidAsyncDraft04");
__createBinding(exports, draft_04_1, "isJSONValid", "isJSONValidDraft04");
__createBinding(exports, draft_04_1, "jsonValidatorAsync", "jsonValidatorAsyncDraft04");
__createBinding(exports, draft_04_1, "jsonValidator", "jsonValidatorDraft04");
__createBinding(exports, draft_04_1, "validateJSONAsync", "validateJSONAsyncDraft04");
__createBinding(exports, draft_04_1, "validateJSON", "validateJSONDraft04");
var draft_06_1 = require("./specification/draft-06");
__createBinding(exports, draft_06_1, "isJSONValidAsync", "isJSONValidAsyncDraft06");
__createBinding(exports, draft_06_1, "isJSONValid", "isJSONValidDraft06");
__createBinding(exports, draft_06_1, "jsonValidatorAsync", "jsonValidatorAsyncDraft06");
__createBinding(exports, draft_06_1, "jsonValidator", "jsonValidatorDraft06");
__createBinding(exports, draft_06_1, "validateJSONAsync", "validateJSONAsyncDraft06");
__createBinding(exports, draft_06_1, "validateJSON", "validateJSONDraft06");
var draft_07_1 = require("./specification/draft-07");
__createBinding(exports, draft_07_1, "isJSONValidAsync", "isJSONValidAsyncDraft07");
__createBinding(exports, draft_07_1, "isJSONValid", "isJSONValidDraft07");
__createBinding(exports, draft_07_1, "jsonValidatorAsync", "jsonValidatorAsyncDraft07");
__createBinding(exports, draft_07_1, "jsonValidator", "jsonValidatorDraft07");
__createBinding(exports, draft_07_1, "validateJSONAsync", "validateJSONAsyncDraft07");
__createBinding(exports, draft_07_1, "validateJSON", "validateJSONDraft07");
var draft_2020_12_1 = require("./specification/draft-2020-12");
__createBinding(exports, draft_2020_12_1, "isJSONValidAsync", "isJSONValidAsyncDraft2020_12");
__createBinding(exports, draft_2020_12_1, "isJSONValid", "isJSONValidDraft2020_12");
__createBinding(exports, draft_2020_12_1, "jsonValidatorAsync", "jsonValidatorAsyncDraft2020_12");
__createBinding(exports, draft_2020_12_1, "jsonValidator", "jsonValidatorDraft2020_12");
__createBinding(exports, draft_2020_12_1, "validateJSONAsync", "validateJSONAsyncDraft2020_12");
__createBinding(exports, draft_2020_12_1, "validateJSON", "validateJSONDraft2020_12");
