// @bufbuild/cel/dist/cjs/index.js
"use strict";
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.celEnv = exports.parse = exports.plan = exports.run = exports.isObjectCelType = exports.isCelType = exports.mapType = exports.typeType = exports.listType = exports.CelScalar = exports.objectType = exports.celType = exports.celFromScalar = exports.isCelUint = exports.celUint = exports.celListConcat = exports.isCelList = exports.celList = exports.isCelMap = exports.celMap = exports.celOverload = exports.celFunc = exports.isCelError = exports.celError = void 0;
var error_js_1 = require("./error.js");
Object.defineProperty(exports, "celError", { enumerable: true, get: function () { return error_js_1.celError; } });
Object.defineProperty(exports, "isCelError", { enumerable: true, get: function () { return error_js_1.isCelError; } });
var func_js_1 = require("./func.js");
Object.defineProperty(exports, "celFunc", { enumerable: true, get: function () { return func_js_1.celFunc; } });
Object.defineProperty(exports, "celOverload", { enumerable: true, get: function () { return func_js_1.celOverload; } });
var map_js_1 = require("./map.js");
Object.defineProperty(exports, "celMap", { enumerable: true, get: function () { return map_js_1.celMap; } });
Object.defineProperty(exports, "isCelMap", { enumerable: true, get: function () { return map_js_1.isCelMap; } });
var list_js_1 = require("./list.js");
Object.defineProperty(exports, "celList", { enumerable: true, get: function () { return list_js_1.celList; } });
Object.defineProperty(exports, "isCelList", { enumerable: true, get: function () { return list_js_1.isCelList; } });
Object.defineProperty(exports, "celListConcat", { enumerable: true, get: function () { return list_js_1.celListConcat; } });
var uint_js_1 = require("./uint.js");
Object.defineProperty(exports, "celUint", { enumerable: true, get: function () { return uint_js_1.celUint; } });
Object.defineProperty(exports, "isCelUint", { enumerable: true, get: function () { return uint_js_1.isCelUint; } });
var proto_js_1 = require("./proto.js");
Object.defineProperty(exports, "celFromScalar", { enumerable: true, get: function () { return proto_js_1.celFromScalar; } });
var type_js_1 = require("./type.js");
Object.defineProperty(exports, "celType", { enumerable: true, get: function () { return type_js_1.celType; } });
Object.defineProperty(exports, "objectType", { enumerable: true, get: function () { return type_js_1.objectType; } });
Object.defineProperty(exports, "CelScalar", { enumerable: true, get: function () { return type_js_1.CelScalar; } });
Object.defineProperty(exports, "listType", { enumerable: true, get: function () { return type_js_1.listType; } });
Object.defineProperty(exports, "typeType", { enumerable: true, get: function () { return type_js_1.typeType; } });
Object.defineProperty(exports, "mapType", { enumerable: true, get: function () { return type_js_1.mapType; } });
Object.defineProperty(exports, "isCelType", { enumerable: true, get: function () { return type_js_1.isCelType; } });
Object.defineProperty(exports, "isObjectCelType", { enumerable: true, get: function () { return type_js_1.isObjectCelType; } });
var run_js_1 = require("./run.js");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return run_js_1.run; } });
var plan_js_1 = require("./plan.js");
Object.defineProperty(exports, "plan", { enumerable: true, get: function () { return plan_js_1.plan; } });
var parse_js_1 = require("./parse.js");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parse_js_1.parse; } });
var env_js_1 = require("./env.js");
Object.defineProperty(exports, "celEnv", { enumerable: true, get: function () { return env_js_1.celEnv; } });
