// @bufbuild/cel/dist/cjs/gen/dev/cel/expr/operator_const.js
"use strict";
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUBTRACT = exports.OPT_SELECT = exports.OPT_INDEX = exports.OLD_NOT_STRICTLY_FALSE = exports.OLD_IN = exports.NOT_STRICTLY_FALSE = exports.NOT_EQUALS = exports.NEGATE = exports.MULTIPLY = exports.MODULO = exports.MAP = exports.LOGICAL_OR = exports.LOGICAL_NOT = exports.LOGICAL_AND = exports.LESS_EQUALS = exports.LESS = exports.INDEX = exports.IN = exports.HAS = exports.GREATER_EQUALS = exports.GREATER = exports.FILTER = exports.EXISTS_ONE = exports.EXISTS = exports.EQUALS = exports.DIVIDE = exports.CONDITIONAL = exports.ALL = exports.ADD = void 0;
// This file is generated by buf-yaml. DO NOT EDIT.
// generated from: proto/dev/cel/expr/operator.const.yaml
// TODO convert to enum
// Also see https://github.com/bufbuild/cel-es/blob/main/packages/cel/proto/dev/cel/expr/operator.const.yaml
exports.ADD = "_+_";
exports.ALL = "all";
exports.CONDITIONAL = "_?_:_";
exports.DIVIDE = "_/_";
exports.EQUALS = "_==_";
exports.EXISTS = "exists";
exports.EXISTS_ONE = "exists_one";
exports.FILTER = "filter";
exports.GREATER = "_>_";
exports.GREATER_EQUALS = "_>=_";
exports.HAS = "has";
exports.IN = "@in";
exports.INDEX = "_[_]";
exports.LESS = "_<_";
exports.LESS_EQUALS = "_<=_";
exports.LOGICAL_AND = "_&&_";
exports.LOGICAL_NOT = "!_";
exports.LOGICAL_OR = "_||_";
exports.MAP = "map";
exports.MODULO = "_%_";
exports.MULTIPLY = "_*_";
exports.NEGATE = "-_";
exports.NOT_EQUALS = "_!=_";
exports.NOT_STRICTLY_FALSE = "@not_strictly_false";
exports.OLD_IN = "_in_";
exports.OLD_NOT_STRICTLY_FALSE = "__not_strictly_false__";
exports.OPT_INDEX = "_[?_]";
exports.OPT_SELECT = "_?._";
exports.SUBTRACT = "_-_";
