// @bufbuild/cel/dist/cjs/gen/dev/cel/expr/overload_const.js
"use strict";
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GREATER_EQUALS_STRING = exports.GREATER_EQUALS_INT64_UINT64 = exports.GREATER_EQUALS_INT64_DOUBLE = exports.GREATER_EQUALS_INT64 = exports.GREATER_EQUALS_DURATION = exports.GREATER_EQUALS_DOUBLE_UINT64 = exports.GREATER_EQUALS_DOUBLE_INT64 = exports.GREATER_EQUALS_DOUBLE = exports.GREATER_EQUALS_BYTES = exports.GREATER_EQUALS_BOOL = exports.GREATER_DURATION = exports.GREATER_DOUBLE_UINT64 = exports.GREATER_DOUBLE_INT64 = exports.GREATER_DOUBLE = exports.GREATER_BYTES = exports.GREATER_BOOL = exports.EQUALS = exports.ENDS_WITH_STRING = exports.ENDS_WITH = exports.DURATION_TO_STRING = exports.DURATION_TO_SECONDS = exports.DURATION_TO_MINUTES = exports.DURATION_TO_MILLISECONDS = exports.DURATION_TO_INT = exports.DURATION_TO_HOURS = exports.DURATION_TO_DURATION = exports.DOUBLE_TO_UINT = exports.DOUBLE_TO_STRING = exports.DOUBLE_TO_INT = exports.DOUBLE_TO_DOUBLE = exports.DIVIDE_UINT64 = exports.DIVIDE_INT64 = exports.DIVIDE_DOUBLE = exports.DEPRECATED_IN = exports.CONTAINS_STRING = exports.CONTAINS = exports.CONDITIONAL = exports.BYTES_TO_STRING = exports.BYTES_TO_BYTES = exports.BOOL_TO_STRING = exports.BOOL_TO_BOOL = exports.ADD_UINT64 = exports.ADD_TIMESTAMP_DURATION = exports.ADD_STRING = exports.ADD_LIST = exports.ADD_INT64 = exports.ADD_DURATION_TIMESTAMP = exports.ADD_DURATION_DURATION = exports.ADD_DOUBLE = exports.ADD_BYTES = void 0;
exports.LESS_STRING = exports.LESS_INT64_UINT64 = exports.LESS_INT64_DOUBLE = exports.LESS_INT64 = exports.LESS_EQUALS_UINT64_INT64 = exports.LESS_EQUALS_UINT64_DOUBLE = exports.LESS_EQUALS_UINT64 = exports.LESS_EQUALS_TIMESTAMP = exports.LESS_EQUALS_STRING = exports.LESS_EQUALS_INT64_UINT64 = exports.LESS_EQUALS_INT64_DOUBLE = exports.LESS_EQUALS_INT64 = exports.LESS_EQUALS_DURATION = exports.LESS_EQUALS_DOUBLE_UINT64 = exports.LESS_EQUALS_DOUBLE_INT64 = exports.LESS_EQUALS_DOUBLE = exports.LESS_EQUALS_BYTES = exports.LESS_EQUALS_BOOL = exports.LESS_DURATION = exports.LESS_DOUBLE_UINT64 = exports.LESS_DOUBLE_INT64 = exports.LESS_DOUBLE = exports.LESS_BYTES = exports.LESS_BOOL = exports.ITERATOR = exports.IN_MESSAGE = exports.IN_MAP = exports.IN_LIST = exports.INT_TO_UINT = exports.INT_TO_TIMESTAMP = exports.INT_TO_STRING = exports.INT_TO_INT = exports.INT_TO_DURATION = exports.INT_TO_DOUBLE = exports.INDEX_MESSAGE = exports.INDEX_MAP = exports.INDEX_LIST = exports.HAS_NEXT = exports.GREATER_UINT64_INT64 = exports.GREATER_UINT64_DOUBLE = exports.GREATER_UINT64 = exports.GREATER_TIMESTAMP = exports.GREATER_STRING = exports.GREATER_INT64_UINT64 = exports.GREATER_INT64_DOUBLE = exports.GREATER_INT64 = exports.GREATER_EQUALS_UINT64_INT64 = exports.GREATER_EQUALS_UINT64_DOUBLE = exports.GREATER_EQUALS_UINT64 = exports.GREATER_EQUALS_TIMESTAMP = void 0;
exports.TIMESTAMP_TO_DAY_OF_WEEK_WITH_TZ = exports.TIMESTAMP_TO_DAY_OF_WEEK = exports.TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED_WITH_TZ = exports.TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED = exports.TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED_WITH_TZ = exports.TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED = exports.SUBTRACT_UINT64 = exports.SUBTRACT_TIMESTAMP_TIMESTAMP = exports.SUBTRACT_TIMESTAMP_DURATION = exports.SUBTRACT_INT64 = exports.SUBTRACT_DURATION_DURATION = exports.SUBTRACT_DOUBLE = exports.STRING_TO_UINT = exports.STRING_TO_TIMESTAMP = exports.STRING_TO_STRING = exports.STRING_TO_INT = exports.STRING_TO_DURATION = exports.STRING_TO_DOUBLE = exports.STRING_TO_BYTES = exports.STRING_TO_BOOL = exports.STARTS_WITH_STRING = exports.STARTS_WITH = exports.SIZE_STRING_INST = exports.SIZE_STRING = exports.SIZE_MAP_INST = exports.SIZE_MAP = exports.SIZE_LIST_INST = exports.SIZE_LIST = exports.SIZE_BYTES_INST = exports.SIZE_BYTES = exports.SIZE = exports.NOT_STRICTLY_FALSE = exports.NOT_EQUALS = exports.NEXT = exports.NEGATE_INT64 = exports.NEGATE_DOUBLE = exports.MULTIPLY_UINT64 = exports.MULTIPLY_INT64 = exports.MULTIPLY_DOUBLE = exports.MODULO_UINT64 = exports.MODULO_INT64 = exports.MATCHES_STRING = exports.MATCHES = exports.LOGICAL_OR = exports.LOGICAL_NOT = exports.LOGICAL_AND = exports.LESS_UINT64_INT64 = exports.LESS_UINT64_DOUBLE = exports.LESS_UINT64 = exports.LESS_TIMESTAMP = void 0;
exports.UINT_TO_UINT = exports.UINT_TO_STRING = exports.UINT_TO_INT = exports.UINT_TO_DOUBLE = exports.TYPE_CONVERT_UINT = exports.TYPE_CONVERT_TYPE = exports.TYPE_CONVERT_TIMESTAMP = exports.TYPE_CONVERT_STRING = exports.TYPE_CONVERT_INT = exports.TYPE_CONVERT_DYN = exports.TYPE_CONVERT_DURATION = exports.TYPE_CONVERT_DOUBLE = exports.TYPE_CONVERT_BYTES = exports.TYPE_CONVERT_BOOL = exports.TO_DYN = exports.TIME_GET_SECONDS = exports.TIME_GET_MONTH = exports.TIME_GET_MINUTES = exports.TIME_GET_MILLISECONDS = exports.TIME_GET_HOURS = exports.TIME_GET_FULL_YEAR = exports.TIME_GET_DAY_OF_YEAR = exports.TIME_GET_DAY_OF_WEEK = exports.TIME_GET_DAY_OF_MONTH = exports.TIME_GET_DATE = exports.TIMESTAMP_TO_YEAR_WITH_TZ = exports.TIMESTAMP_TO_YEAR = exports.TIMESTAMP_TO_TIMESTAMP = exports.TIMESTAMP_TO_STRING = exports.TIMESTAMP_TO_SECONDS_WITH_TZ = exports.TIMESTAMP_TO_SECONDS = exports.TIMESTAMP_TO_MONTH_WITH_TZ = exports.TIMESTAMP_TO_MONTH = exports.TIMESTAMP_TO_MINUTES_WITH_TZ = exports.TIMESTAMP_TO_MINUTES = exports.TIMESTAMP_TO_MILLISECONDS_WITH_TZ = exports.TIMESTAMP_TO_MILLISECONDS = exports.TIMESTAMP_TO_INT = exports.TIMESTAMP_TO_HOURS_WITH_TZ = exports.TIMESTAMP_TO_HOURS = exports.TIMESTAMP_TO_DAY_OF_YEAR_WITH_TZ = exports.TIMESTAMP_TO_DAY_OF_YEAR = void 0;
// This file is generated by buf-yaml. DO NOT EDIT.
// generated from: proto/dev/cel/expr/overload.const.yaml
// TODO convert to enum
// Also see https://github.com/bufbuild/cel-es/blob/main/packages/cel/proto/dev/cel/expr/operator.const.yaml
exports.ADD_BYTES = "add_bytes";
exports.ADD_DOUBLE = "add_double";
exports.ADD_DURATION_DURATION = "add_duration_duration";
exports.ADD_DURATION_TIMESTAMP = "add_duration_timestamp";
exports.ADD_INT64 = "add_int64";
exports.ADD_LIST = "add_list";
exports.ADD_STRING = "add_string";
exports.ADD_TIMESTAMP_DURATION = "add_timestamp_duration";
exports.ADD_UINT64 = "add_uint64";
exports.BOOL_TO_BOOL = "bool_to_bool";
exports.BOOL_TO_STRING = "bool_to_string";
exports.BYTES_TO_BYTES = "bytes_to_bytes";
exports.BYTES_TO_STRING = "bytes_to_string";
exports.CONDITIONAL = "conditional";
exports.CONTAINS = "contains";
exports.CONTAINS_STRING = "contains_string";
exports.DEPRECATED_IN = "in";
exports.DIVIDE_DOUBLE = "divide_double";
exports.DIVIDE_INT64 = "divide_int64";
exports.DIVIDE_UINT64 = "divide_uint64";
exports.DOUBLE_TO_DOUBLE = "double_to_double";
exports.DOUBLE_TO_INT = "double_to_int64";
exports.DOUBLE_TO_STRING = "double_to_string";
exports.DOUBLE_TO_UINT = "double_to_uint64";
exports.DURATION_TO_DURATION = "duration_to_duration";
exports.DURATION_TO_HOURS = "duration_to_hours";
exports.DURATION_TO_INT = "duration_to_int64";
exports.DURATION_TO_MILLISECONDS = "duration_to_milliseconds";
exports.DURATION_TO_MINUTES = "duration_to_minutes";
exports.DURATION_TO_SECONDS = "duration_to_seconds";
exports.DURATION_TO_STRING = "duration_to_string";
exports.ENDS_WITH = "endsWith";
exports.ENDS_WITH_STRING = "ends_with_string";
exports.EQUALS = "equals";
exports.GREATER_BOOL = "greater_bool";
exports.GREATER_BYTES = "greater_bytes";
exports.GREATER_DOUBLE = "greater_double";
exports.GREATER_DOUBLE_INT64 = "greater_double_int64";
exports.GREATER_DOUBLE_UINT64 = "greater_double_uint64";
exports.GREATER_DURATION = "greater_duration";
exports.GREATER_EQUALS_BOOL = "greater_equals_bool";
exports.GREATER_EQUALS_BYTES = "greater_equals_bytes";
exports.GREATER_EQUALS_DOUBLE = "greater_equals_double";
exports.GREATER_EQUALS_DOUBLE_INT64 = "greater_equals_double_int64";
exports.GREATER_EQUALS_DOUBLE_UINT64 = "greater_equals_double_uint64";
exports.GREATER_EQUALS_DURATION = "greater_equals_duration";
exports.GREATER_EQUALS_INT64 = "greater_equals_int64";
exports.GREATER_EQUALS_INT64_DOUBLE = "greater_equals_int64_double";
exports.GREATER_EQUALS_INT64_UINT64 = "greater_equals_int64_uint64";
exports.GREATER_EQUALS_STRING = "greater_equals_string";
exports.GREATER_EQUALS_TIMESTAMP = "greater_equals_timestamp";
exports.GREATER_EQUALS_UINT64 = "greater_equals_uint64";
exports.GREATER_EQUALS_UINT64_DOUBLE = "greater_equals_uint64_double";
exports.GREATER_EQUALS_UINT64_INT64 = "greater_equals_uint64_int64";
exports.GREATER_INT64 = "greater_int64";
exports.GREATER_INT64_DOUBLE = "greater_int64_double";
exports.GREATER_INT64_UINT64 = "greater_int64_uint64";
exports.GREATER_STRING = "greater_string";
exports.GREATER_TIMESTAMP = "greater_timestamp";
exports.GREATER_UINT64 = "greater_uint64";
exports.GREATER_UINT64_DOUBLE = "greater_uint64_double";
exports.GREATER_UINT64_INT64 = "greater_uint64_int64";
exports.HAS_NEXT = "@hasNext";
exports.INDEX_LIST = "index_list";
exports.INDEX_MAP = "index_map";
exports.INDEX_MESSAGE = "index_message";
exports.INT_TO_DOUBLE = "int64_to_double";
exports.INT_TO_DURATION = "int64_to_duration";
exports.INT_TO_INT = "int64_to_int64";
exports.INT_TO_STRING = "int64_to_string";
exports.INT_TO_TIMESTAMP = "int64_to_timestamp";
exports.INT_TO_UINT = "int64_to_uint64";
exports.IN_LIST = "in_list";
exports.IN_MAP = "in_map";
exports.IN_MESSAGE = "in_message";
exports.ITERATOR = "@iterator";
exports.LESS_BOOL = "less_bool";
exports.LESS_BYTES = "less_bytes";
exports.LESS_DOUBLE = "less_double";
exports.LESS_DOUBLE_INT64 = "less_double_int64";
exports.LESS_DOUBLE_UINT64 = "less_double_uint64";
exports.LESS_DURATION = "less_duration";
exports.LESS_EQUALS_BOOL = "less_equals_bool";
exports.LESS_EQUALS_BYTES = "less_equals_bytes";
exports.LESS_EQUALS_DOUBLE = "less_equals_double";
exports.LESS_EQUALS_DOUBLE_INT64 = "less_equals_double_int64";
exports.LESS_EQUALS_DOUBLE_UINT64 = "less_equals_double_uint64";
exports.LESS_EQUALS_DURATION = "less_equals_duration";
exports.LESS_EQUALS_INT64 = "less_equals_int64";
exports.LESS_EQUALS_INT64_DOUBLE = "less_equals_int64_double";
exports.LESS_EQUALS_INT64_UINT64 = "less_equals_int64_uint64";
exports.LESS_EQUALS_STRING = "less_equals_string";
exports.LESS_EQUALS_TIMESTAMP = "less_equals_timestamp";
exports.LESS_EQUALS_UINT64 = "less_equals_uint64";
exports.LESS_EQUALS_UINT64_DOUBLE = "less_equals_uint64_double";
exports.LESS_EQUALS_UINT64_INT64 = "less_equals_uint64_int64";
exports.LESS_INT64 = "less_int64";
exports.LESS_INT64_DOUBLE = "less_int64_double";
exports.LESS_INT64_UINT64 = "less_int64_uint64";
exports.LESS_STRING = "less_string";
exports.LESS_TIMESTAMP = "less_timestamp";
exports.LESS_UINT64 = "less_uint64";
exports.LESS_UINT64_DOUBLE = "less_uint64_double";
exports.LESS_UINT64_INT64 = "less_uint64_int64";
exports.LOGICAL_AND = "logical_and";
exports.LOGICAL_NOT = "logical_not";
exports.LOGICAL_OR = "logical_or";
exports.MATCHES = "matches";
exports.MATCHES_STRING = "matches_string";
exports.MODULO_INT64 = "modulo_int64";
exports.MODULO_UINT64 = "modulo_uint64";
exports.MULTIPLY_DOUBLE = "multiply_double";
exports.MULTIPLY_INT64 = "multiply_int64";
exports.MULTIPLY_UINT64 = "multiply_uint64";
exports.NEGATE_DOUBLE = "negate_double";
exports.NEGATE_INT64 = "negate_int64";
exports.NEXT = "@next";
exports.NOT_EQUALS = "not_equals";
exports.NOT_STRICTLY_FALSE = "not_strictly_false";
exports.SIZE = "size";
exports.SIZE_BYTES = "size_bytes";
exports.SIZE_BYTES_INST = "bytes_size";
exports.SIZE_LIST = "size_list";
exports.SIZE_LIST_INST = "list_size";
exports.SIZE_MAP = "size_map";
exports.SIZE_MAP_INST = "map_size";
exports.SIZE_STRING = "size_string";
exports.SIZE_STRING_INST = "string_size";
exports.STARTS_WITH = "startsWith";
exports.STARTS_WITH_STRING = "starts_with_string";
exports.STRING_TO_BOOL = "string_to_bool";
exports.STRING_TO_BYTES = "string_to_bytes";
exports.STRING_TO_DOUBLE = "string_to_double";
exports.STRING_TO_DURATION = "string_to_duration";
exports.STRING_TO_INT = "string_to_int64";
exports.STRING_TO_STRING = "string_to_string";
exports.STRING_TO_TIMESTAMP = "string_to_timestamp";
exports.STRING_TO_UINT = "string_to_uint64";
exports.SUBTRACT_DOUBLE = "subtract_double";
exports.SUBTRACT_DURATION_DURATION = "subtract_duration_duration";
exports.SUBTRACT_INT64 = "subtract_int64";
exports.SUBTRACT_TIMESTAMP_DURATION = "subtract_timestamp_duration";
exports.SUBTRACT_TIMESTAMP_TIMESTAMP = "subtract_timestamp_timestamp";
exports.SUBTRACT_UINT64 = "subtract_uint64";
exports.TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED = "timestamp_to_day_of_month_1_based";
exports.TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED_WITH_TZ = "timestamp_to_day_of_month_1_based_with_tz";
exports.TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED = "timestamp_to_day_of_month";
exports.TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED_WITH_TZ = "timestamp_to_day_of_month_with_tz";
exports.TIMESTAMP_TO_DAY_OF_WEEK = "timestamp_to_day_of_week";
exports.TIMESTAMP_TO_DAY_OF_WEEK_WITH_TZ = "timestamp_to_day_of_week_with_tz";
exports.TIMESTAMP_TO_DAY_OF_YEAR = "timestamp_to_day_of_year";
exports.TIMESTAMP_TO_DAY_OF_YEAR_WITH_TZ = "timestamp_to_day_of_year_with_tz";
exports.TIMESTAMP_TO_HOURS = "timestamp_to_hours";
exports.TIMESTAMP_TO_HOURS_WITH_TZ = "timestamp_to_hours_with_tz";
exports.TIMESTAMP_TO_INT = "timestamp_to_int64";
exports.TIMESTAMP_TO_MILLISECONDS = "timestamp_to_milliseconds";
exports.TIMESTAMP_TO_MILLISECONDS_WITH_TZ = "timestamp_to_milliseconds_with_tz";
exports.TIMESTAMP_TO_MINUTES = "timestamp_to_minutes";
exports.TIMESTAMP_TO_MINUTES_WITH_TZ = "timestamp_to_minutes_with_tz";
exports.TIMESTAMP_TO_MONTH = "timestamp_to_month";
exports.TIMESTAMP_TO_MONTH_WITH_TZ = "timestamp_to_month_with_tz";
exports.TIMESTAMP_TO_SECONDS = "timestamp_to_seconds";
exports.TIMESTAMP_TO_SECONDS_WITH_TZ = "timestamp_to_seconds_tz";
exports.TIMESTAMP_TO_STRING = "timestamp_to_string";
exports.TIMESTAMP_TO_TIMESTAMP = "timestamp_to_timestamp";
exports.TIMESTAMP_TO_YEAR = "timestamp_to_year";
exports.TIMESTAMP_TO_YEAR_WITH_TZ = "timestamp_to_year_with_tz";
exports.TIME_GET_DATE = "getDate";
exports.TIME_GET_DAY_OF_MONTH = "getDayOfMonth";
exports.TIME_GET_DAY_OF_WEEK = "getDayOfWeek";
exports.TIME_GET_DAY_OF_YEAR = "getDayOfYear";
exports.TIME_GET_FULL_YEAR = "getFullYear";
exports.TIME_GET_HOURS = "getHours";
exports.TIME_GET_MILLISECONDS = "getMilliseconds";
exports.TIME_GET_MINUTES = "getMinutes";
exports.TIME_GET_MONTH = "getMonth";
exports.TIME_GET_SECONDS = "getSeconds";
exports.TO_DYN = "to_dyn";
exports.TYPE_CONVERT_BOOL = "bool";
exports.TYPE_CONVERT_BYTES = "bytes";
exports.TYPE_CONVERT_DOUBLE = "double";
exports.TYPE_CONVERT_DURATION = "duration";
exports.TYPE_CONVERT_DYN = "dyn";
exports.TYPE_CONVERT_INT = "int";
exports.TYPE_CONVERT_STRING = "string";
exports.TYPE_CONVERT_TIMESTAMP = "timestamp";
exports.TYPE_CONVERT_TYPE = "type";
exports.TYPE_CONVERT_UINT = "uint";
exports.UINT_TO_DOUBLE = "uint64_to_double";
exports.UINT_TO_INT = "uint64_to_int64";
exports.UINT_TO_STRING = "uint64_to_string";
exports.UINT_TO_UINT = "uint64_to_uint64";
