// @confluentinc/schemaregistry/dist/oauth/static-token-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._StaticTokenProviderBuilder = void 0;
const abstract_oauth_client_1 = require("./abstract-oauth-client");
class StaticTokenProviderBuilder extends abstract_oauth_client_1._AbstractBearerTokenProviderBuilder {
    constructor(bearerAuthCredentials) {
        super(bearerAuthCredentials);
    }
    validate() {
        super.validate();
        if (!this.bearerAuthCredentials.token) {
            throw new Error('Bearer token not provided');
        }
    }
    build() {
        this.validate();
        return new StaticTokenProvider(this.bearerAuthCredentials);
    }
}
exports._StaticTokenProviderBuilder = StaticTokenProviderBuilder;
class StaticTokenProvider extends abstract_oauth_client_1._AbstractOauthTokenProvider {
    constructor(bearerAuthCredentials) {
        super(bearerAuthCredentials);
        this.token = bearerAuthCredentials.token;
    }
    getAccessToken() {
        return Promise.resolve(this.token);
    }
    tokenExpired() {
        return false;
    }
}
