// @criteria/json-schema-validation/dist/specification/draft-04/vocabularies/validation/additionalItemsValidator.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.formatMessage = exports.additionalItemsValidator = void 0;
var formatList_1 = require("../../../../util/formatList");
var isJSONArray_1 = require("../../../../util/isJSONArray");
function additionalItemsValidator(schema, schemaPath, context) {
    var _a;
    if (!('additionalItems' in schema)) {
        return null;
    }
    // if "items" is not present, or its value is an object, validation
    // of the instance always succeeds, regardless of the value of
    // "additionalItems";
    if (!('items' in schema) || !Array.isArray(schema['items'])) {
        return null;
    }
    var additionalItems = schema['additionalItems'];
    var itemsCount = ((_a = schema['items']) !== null && _a !== void 0 ? _a : []).length;
    var outputFormat = context.outputFormat;
    var failFast = context.failFast;
    var schemaLocation = schemaPath.join('');
    // short-cut
    if (additionalItems === true) {
        return null; // TODO: what about annotations
    }
    if (additionalItems === false) {
        return function (instance, instanceLocation, annotationResults) {
            if (!(0, isJSONArray_1.isJSONArray)(instance)) {
                return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
            }
            var valid = instance.length <= itemsCount;
            if (valid) {
                return {
                    valid: true,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'additionalItems',
                    instanceLocation: instanceLocation,
                    annotationResults: {
                        additionalItems: true // TODO: only true if actually applied to additional items
                    }
                };
            }
            else {
                if (outputFormat === 'flag') {
                    return { valid: false };
                }
                else {
                    var invalidIndices = Array.from({ length: instance.length - itemsCount }, function (_, i) { return itemsCount + i; });
                    return {
                        valid: false,
                        schemaLocation: schemaLocation,
                        schemaKeyword: 'additionalItems',
                        instanceLocation: instanceLocation,
                        message: formatMessage(null, invalidIndices)
                    };
                }
            }
        };
    }
    var validator = context.validatorForSchema(additionalItems, __spreadArray(__spreadArray([], __read(schemaPath), false), ['/additionalItems'], false));
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONArray_1.isJSONArray)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        var invalidIndices = [];
        var errors = [];
        for (var i = itemsCount; i < instance.length; i++) {
            var output = validator(instance[i], "".concat(instanceLocation, "/").concat(i));
            if (!output.valid && failFast) {
                if (outputFormat === 'flag') {
                    return { valid: false };
                }
                else {
                    return {
                        valid: false,
                        schemaLocation: schemaLocation,
                        schemaKeyword: 'additionalItems',
                        instanceLocation: instanceLocation,
                        message: formatMessage([output], [i])
                    };
                }
            }
            if (output.valid) {
                // outputs.push(output)
            }
            else {
                invalidIndices.push(i);
                errors.push(output);
            }
        }
        if (errors.length === 0) {
            return {
                valid: true,
                schemaLocation: schemaLocation,
                schemaKeyword: 'additionalItems',
                instanceLocation: instanceLocation,
                annotationResults: {
                    additionalItems: true // TODO: only true if actually applied to additional items
                }
            };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'additionalItems',
                    instanceLocation: instanceLocation,
                    message: formatMessage(errors, invalidIndices),
                    errors: errors
                };
            }
        }
    };
}
exports.additionalItemsValidator = additionalItemsValidator;
function formatMessage(errors, invalidIndices) {
    var message;
    if (invalidIndices.length === 1) {
        message = "has an invalid item at position ".concat(invalidIndices[0]);
    }
    else {
        message = "has invalid items at positions ".concat((0, formatList_1.formatList)(invalidIndices.map(function (invalidIndex) { return "".concat(invalidIndex); }), 'and'));
    }
    if (errors !== null) {
        message += " (".concat(errors.map(function (error) { return error.message; }).join('; '), ")");
    }
    return message;
}
exports.formatMessage = formatMessage;
