// @criteria/json-schema-validation/dist/specification/draft-04/vocabularies/validation/allOfValidator.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.formatMessage = exports.allOfValidator = void 0;
var formatList_1 = require("../../../../util/formatList");
var reduceAnnotationResults_1 = require("../reduceAnnotationResults");
function allOfValidator(schema, schemaPath, context) {
    if (!('allOf' in schema)) {
        return null;
    }
    var allOf = schema['allOf'];
    var validators = allOf.map(function (subschema, i) { return context.validatorForSchema(subschema, __spreadArray(__spreadArray([], __read(schemaPath), false), ["/allOf/".concat(i)], false)); });
    var outputFormat = context.outputFormat;
    var failFast = context.failFast;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        var validOutputs = [];
        var errors = [];
        for (var i = 0; i < validators.length; i++) {
            var validator = validators[i];
            var output = validator(instance, instanceLocation);
            if (output.valid) {
                validOutputs.push(output);
            }
            else {
                if (failFast) {
                    if (outputFormat === 'flag') {
                        return { valid: false };
                    }
                    else {
                        return {
                            valid: false,
                            schemaLocation: schemaLocation,
                            schemaKeyword: 'allOf',
                            instanceLocation: instanceLocation,
                            message: formatMessage([output]),
                            errors: [output]
                        };
                    }
                }
                errors.push(output);
            }
        }
        if (errors.length === 0) {
            return {
                valid: true,
                schemaLocation: schemaLocation,
                schemaKeyword: 'allOf',
                instanceLocation: instanceLocation,
                annotationResults: validOutputs
                    .map(function (output) { var _a; return (_a = output.annotationResults) !== null && _a !== void 0 ? _a : {}; })
                    .reduce(reduceAnnotationResults_1.reduceAnnotationResults, {})
            };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'allOf',
                    instanceLocation: instanceLocation,
                    message: formatMessage(errors),
                    errors: errors
                };
            }
        }
    };
}
exports.allOfValidator = allOfValidator;
function formatMessage(errors) {
    return (0, formatList_1.formatList)(errors.map(function (error) { return error.message; }), 'and');
}
exports.formatMessage = formatMessage;
