// @criteria/json-schema-validation/dist/specification/draft-04/vocabularies/validation/index.js
"use strict";
exports.__esModule = true;
exports.validationValidators = void 0;
var additionalItemsValidator_1 = require("./additionalItemsValidator");
var additionalPropertiesValidator_1 = require("./additionalPropertiesValidator");
var allOfValidator_1 = require("./allOfValidator");
var anyOfValidator_1 = require("./anyOfValidator");
var constValidator_1 = require("./constValidator");
var dependenciesValidator_1 = require("./dependenciesValidator");
var enumValidator_1 = require("./enumValidator");
var formatValidator_1 = require("./formatValidator");
var itemsValidator_1 = require("./itemsValidator");
var maxItemsValidator_1 = require("./maxItemsValidator");
var maxLengthValidator_1 = require("./maxLengthValidator");
var maxPropertiesValidator_1 = require("./maxPropertiesValidator");
var maximumValidator_1 = require("./maximumValidator");
var minItemsValidator_1 = require("./minItemsValidator");
var minLengthValidator_1 = require("./minLengthValidator");
var minPropertiesValidator_1 = require("./minPropertiesValidator");
var minimumValidator_1 = require("./minimumValidator");
var multipleOfValidator_1 = require("./multipleOfValidator");
var notValidator_1 = require("./notValidator");
var oneOfValidator_1 = require("./oneOfValidator");
var patternPropertiesValidator_1 = require("./patternPropertiesValidator");
var patternValidator_1 = require("./patternValidator");
var propertiesValidator_1 = require("./propertiesValidator");
var requiredValidator_1 = require("./requiredValidator");
var typeValidator_1 = require("./typeValidator");
var uniqueItemsValidator_1 = require("./uniqueItemsValidator");
exports.validationValidators = {
    type: typeValidator_1.typeValidator,
    "enum": enumValidator_1.enumValidator,
    "const": constValidator_1.constValidator,
    pattern: patternValidator_1.patternValidator,
    minLength: minLengthValidator_1.minLengthValidator,
    maxLength: maxLengthValidator_1.maxLengthValidator,
    multipleOf: multipleOfValidator_1.multipleOfValidator,
    maximum: maximumValidator_1.maximumValidator,
    minimum: minimumValidator_1.minimumValidator,
    dependencies: dependenciesValidator_1.dependenciesValidator,
    maxProperties: maxPropertiesValidator_1.maxPropertiesValidator,
    minProperties: minPropertiesValidator_1.minPropertiesValidator,
    required: requiredValidator_1.requiredValidator,
    items: itemsValidator_1.itemsValidator,
    additionalItems: additionalItemsValidator_1.additionalItemsValidator,
    maxItems: maxItemsValidator_1.maxItemsValidator,
    minItems: minItemsValidator_1.minItemsValidator,
    uniqueItems: uniqueItemsValidator_1.uniqueItemsValidator,
    properties: propertiesValidator_1.propertiesValidator,
    patternProperties: patternPropertiesValidator_1.patternPropertiesValidator,
    additionalProperties: additionalPropertiesValidator_1.additionalPropertiesValidator,
    allOf: allOfValidator_1.allOfValidator,
    anyOf: anyOfValidator_1.anyOfValidator,
    oneOf: oneOfValidator_1.oneOfValidator,
    not: notValidator_1.notValidator,
    format: formatValidator_1.formatValidator
};
