// @criteria/json-schema-validation/dist/specification/draft-04/vocabularies/validation/maxPropertiesValidator.js
"use strict";
exports.__esModule = true;
exports.maxPropertiesValidator = void 0;
var isJSONObject_1 = require("../../../../util/isJSONObject");
function maxPropertiesValidator(schema, schemaPath, context) {
    if (!('maxProperties' in schema)) {
        return null;
    }
    var maxProperties = schema['maxProperties'];
    var outputFormat = context.outputFormat;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONObject_1.isJSONObject)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        var count = Object.keys(instance).length;
        if (count <= maxProperties) {
            return { valid: true, schemaLocation: schemaLocation, schemaKeyword: 'maxProperties', instanceLocation: instanceLocation };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'maxProperties',
                    instanceLocation: instanceLocation,
                    message: maxProperties === 1
                        ? "should have up to 1 property but has ".concat(count, " instead")
                        : "should have up to ".concat(maxProperties, " properties but has ").concat(count, " instead")
                };
            }
        }
    };
}
exports.maxPropertiesValidator = maxPropertiesValidator;
