// @criteria/json-schema-validation/dist/specification/draft-04/vocabularies/validation/patternPropertiesValidator.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
exports.__esModule = true;
exports.formatMessage = exports.patternPropertiesValidator = void 0;
var json_pointer_1 = require("@criteria/json-pointer");
var formatList_1 = require("../../../../util/formatList");
var isJSONObject_1 = require("../../../../util/isJSONObject");
function patternPropertiesValidator(schema, schemaPath, context) {
    if (!('patternProperties' in schema)) {
        return null;
    }
    var patternProperties = schema['patternProperties'];
    var patternValidators = Object.keys(patternProperties).map(function (pattern) {
        var regexp = new RegExp(pattern);
        var subschema = patternProperties[pattern];
        var subschemaValidator = context.validatorForSchema(subschema, __spreadArray(__spreadArray([], __read(schemaPath), false), [
            "/patternProperties/".concat((0, json_pointer_1.escapeReferenceToken)(pattern))
        ], false));
        return [regexp, subschemaValidator];
    });
    var outputFormat = context.outputFormat;
    var failFast = context.failFast;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        var e_1, _a;
        if (!(0, isJSONObject_1.isJSONObject)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        var validKeys = new Set();
        var invalidPropertyNames = [];
        var errors = [];
        var allPropertyNames = Object.keys(instance);
        var _loop_1 = function (regexp, validator) {
            var e_2, _c;
            var propertyNames = allPropertyNames.filter(function (propertyName) { return regexp.test(propertyName); });
            try {
                for (var propertyNames_1 = (e_2 = void 0, __values(propertyNames)), propertyNames_1_1 = propertyNames_1.next(); !propertyNames_1_1.done; propertyNames_1_1 = propertyNames_1.next()) {
                    var propertyName = propertyNames_1_1.value;
                    var output = validator(instance[propertyName], "".concat(instanceLocation, "/").concat((0, json_pointer_1.escapeReferenceToken)(propertyName)));
                    if (output.valid) {
                        validKeys.add(propertyName);
                    }
                    else {
                        if (failFast) {
                            if (outputFormat === 'flag') {
                                return { value: { valid: false } };
                            }
                            else {
                                return { value: {
                                        valid: false,
                                        schemaLocation: schemaLocation,
                                        schemaKeyword: 'patternProperties',
                                        instanceLocation: instanceLocation,
                                        message: formatMessage([output], [propertyName]),
                                        errors: [output]
                                    } };
                            }
                        }
                        errors.push(output);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (propertyNames_1_1 && !propertyNames_1_1.done && (_c = propertyNames_1["return"])) _c.call(propertyNames_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        };
        try {
            for (var patternValidators_1 = __values(patternValidators), patternValidators_1_1 = patternValidators_1.next(); !patternValidators_1_1.done; patternValidators_1_1 = patternValidators_1.next()) {
                var _b = __read(patternValidators_1_1.value, 2), regexp = _b[0], validator = _b[1];
                var state_1 = _loop_1(regexp, validator);
                if (typeof state_1 === "object")
                    return state_1.value;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (patternValidators_1_1 && !patternValidators_1_1.done && (_a = patternValidators_1["return"])) _a.call(patternValidators_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (errors.length === 0) {
            return {
                valid: true,
                schemaLocation: schemaLocation,
                schemaKeyword: 'patternProperties',
                instanceLocation: instanceLocation,
                annotationResults: {
                    patternProperties: Array.from(validKeys)
                }
            };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'patternProperties',
                    instanceLocation: instanceLocation,
                    message: formatMessage(errors, invalidPropertyNames),
                    errors: errors
                };
            }
        }
    };
}
exports.patternPropertiesValidator = patternPropertiesValidator;
function formatMessage(errors, invalidPropertyNames) {
    var message;
    if (invalidPropertyNames.length === 1) {
        message = "has an invalid property '".concat(invalidPropertyNames[0], "'");
    }
    else {
        message = "has invalid properties ".concat((0, formatList_1.formatList)(invalidPropertyNames.map(function (propertyName) { return "'".concat(propertyName, "'"); }), 'and'));
    }
    if (errors !== null) {
        message += " (".concat(errors.map(function (error) { return error.message; }).join('; '), ")");
    }
    return message;
}
exports.formatMessage = formatMessage;
