// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/applicator/index.js
"use strict";
exports.__esModule = true;
exports.applicatorValidators = void 0;
var additionalPropertiesValidator_1 = require("./additionalPropertiesValidator");
var allOfValidator_1 = require("./allOfValidator");
var anyOfValidator_1 = require("./anyOfValidator");
var containsValidator_1 = require("./containsValidator");
var dependentSchemasValidator_1 = require("./dependentSchemasValidator");
var ifValidator_1 = require("./ifValidator");
var itemsValidator_1 = require("./itemsValidator");
var notValidator_1 = require("./notValidator");
var oneOfValidator_1 = require("./oneOfValidator");
var patternPropertiesValidator_1 = require("./patternPropertiesValidator");
var prefixItemsValidator_1 = require("./prefixItemsValidator");
var propertiesValidator_1 = require("./propertiesValidator");
var propertyNamesValidator_1 = require("./propertyNamesValidator");
exports.applicatorValidators = {
    allOf: allOfValidator_1.allOfValidator,
    anyOf: anyOfValidator_1.anyOfValidator,
    oneOf: oneOfValidator_1.oneOfValidator,
    "if": ifValidator_1.ifValidator,
    not: notValidator_1.notValidator,
    properties: propertiesValidator_1.propertiesValidator,
    additionalProperties: additionalPropertiesValidator_1.additionalPropertiesValidator,
    patternProperties: patternPropertiesValidator_1.patternPropertiesValidator,
    dependentSchemas: dependentSchemasValidator_1.dependentSchemasValidator,
    propertyNames: propertyNamesValidator_1.propertyNamesValidator,
    items: itemsValidator_1.itemsValidator,
    prefixItems: prefixItemsValidator_1.prefixItemsValidator,
    contains: containsValidator_1.containsValidator
};
