// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/applicator/oneOfValidator.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
exports.oneOfValidator = void 0;
function oneOfValidator(schema, schemaPath, context) {
    if (!('oneOf' in schema)) {
        return null;
    }
    var oneOf = schema['oneOf'];
    var validators = oneOf.map(function (subschema, i) { return context.validatorForSchema(subschema, __spreadArray(__spreadArray([], __read(schemaPath), false), ["/oneOf/".concat(i)], false)); });
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        var outputs = validators.map(function (validator) { return validator(instance, instanceLocation); });
        var valid = outputs.filter(function (output) { return output.valid; }).length === 1;
        if (valid) {
            var validOutput = outputs.find(function (output) { return output.valid; });
            return {
                valid: true,
                schemaLocation: schemaLocation,
                schemaKeyword: 'oneOf',
                instanceLocation: instanceLocation,
                annotationResults: 'annotationResults' in validOutput ? validOutput.annotationResults : {}
            };
        }
        else {
            var validCount = outputs.filter(function (output) { return output.valid; }).length;
            return {
                valid: false,
                schemaLocation: schemaLocation,
                schemaKeyword: 'oneOf',
                instanceLocation: instanceLocation,
                message: "should validate against exactly one subschema but validates against ".concat(validCount === 0 ? 'none' : validCount)
            };
        }
    };
}
exports.oneOfValidator = oneOfValidator;
