// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/applicator/patternPropertiesValidator.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
exports.__esModule = true;
exports.formatMessage = exports.patternPropertiesValidator = void 0;
var json_pointer_1 = require("@criteria/json-pointer");
var formatList_1 = require("../../../../util/formatList");
var isJSONObject_1 = require("../../../../util/isJSONObject");
function patternPropertiesValidator(schema, schemaPath, context) {
    if (!('patternProperties' in schema)) {
        return null;
    }
    var patternProperties = schema['patternProperties'];
    var patternValidators = Object.keys(patternProperties).map(function (pattern) {
        var regexp = new RegExp(pattern, 'u');
        var subschema = patternProperties[pattern];
        var subschemaValidator = context.validatorForSchema(subschema, __spreadArray(__spreadArray([], __read(schemaPath), false), [
            "/patternProperties/".concat((0, json_pointer_1.escapeReferenceToken)(pattern))
        ], false));
        return [pattern, regexp, subschemaValidator];
    });
    var outputFormat = context.outputFormat;
    var failFast = context.failFast;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        var e_1, _a, e_2, _b;
        if (!(0, isJSONObject_1.isJSONObject)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        var validOutputs = new Map();
        var invalidPropertyNames = [];
        var errors = [];
        try {
            for (var _c = __values(Object.entries(instance)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var _e = __read(_d.value, 2), propertyName = _e[0], propertyValue = _e[1];
                try {
                    for (var patternValidators_1 = (e_2 = void 0, __values(patternValidators)), patternValidators_1_1 = patternValidators_1.next(); !patternValidators_1_1.done; patternValidators_1_1 = patternValidators_1.next()) {
                        var _f = __read(patternValidators_1_1.value, 3), pattern = _f[0], regexp = _f[1], validator = _f[2];
                        // what if multiple patterns match the property?
                        if (!regexp.test(propertyName)) {
                            continue;
                        }
                        var output = validator(propertyValue, "".concat(instanceLocation, "/").concat((0, json_pointer_1.escapeReferenceToken)(propertyName)));
                        if (!output.valid && failFast) {
                            if (outputFormat === 'flag') {
                                return { valid: false };
                            }
                            else {
                                return {
                                    valid: false,
                                    schemaLocation: schemaLocation,
                                    schemaKeyword: 'patternProperties',
                                    instanceLocation: instanceLocation,
                                    message: formatMessage([output], [propertyName]),
                                    errors: [output]
                                };
                            }
                        }
                        if (output.valid) {
                            validOutputs.set(propertyName, output);
                        }
                        else {
                            invalidPropertyNames.push(propertyName);
                            errors.push(output);
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (patternValidators_1_1 && !patternValidators_1_1.done && (_b = patternValidators_1["return"])) _b.call(patternValidators_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                // Property didn't match name or pattern
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c["return"])) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (errors.length === 0) {
            return {
                valid: true,
                schemaLocation: schemaLocation,
                schemaKeyword: 'patternProperties',
                instanceLocation: instanceLocation,
                annotationResults: {
                    patternProperties: Array.from(validOutputs.keys())
                }
            };
        }
        else {
            return {
                valid: false,
                schemaLocation: schemaLocation,
                schemaKeyword: 'patternProperties',
                instanceLocation: instanceLocation,
                message: formatMessage(errors, invalidPropertyNames),
                errors: errors
            };
        }
    };
}
exports.patternPropertiesValidator = patternPropertiesValidator;
function formatMessage(errors, invalidPropertyNames) {
    var message;
    if (invalidPropertyNames.length === 1) {
        message = "has an invalid property ".concat(invalidPropertyNames[0]);
    }
    else {
        message = "has invalid properties ".concat((0, formatList_1.formatList)(invalidPropertyNames.map(function (propertyName) { return "'".concat(propertyName, "'"); }), 'and'));
    }
    if (errors !== null) {
        message += " (".concat(errors.map(function (error) { return error.message; }).join('; '), ")");
    }
    return message;
}
exports.formatMessage = formatMessage;
