// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/validation/exclusiveMinimumValidator.js
"use strict";
exports.__esModule = true;
exports.exclusiveMinimumValidator = void 0;
var format_1 = require("../../../../util/format");
var isJSONNumber_1 = require("../../../../util/isJSONNumber");
function exclusiveMinimumValidator(schema, schemaPath, context) {
    if (!('exclusiveMinimum' in schema)) {
        return null;
    }
    var exclusiveMinimum = schema['exclusiveMinimum'];
    var outputFormat = context.outputFormat;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONNumber_1.isJSONNumber)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        if (instance > exclusiveMinimum) {
            return { valid: true, schemaLocation: schemaLocation, schemaKeyword: 'multipleOf', instanceLocation: instanceLocation };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'exclusiveMinimum',
                    instanceLocation: instanceLocation,
                    message: "should be greater than ".concat(exclusiveMinimum, " but is ").concat((0, format_1.format)(instance), " instead")
                };
            }
        }
    };
}
exports.exclusiveMinimumValidator = exclusiveMinimumValidator;
