// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/validation/index.js
"use strict";
exports.__esModule = true;
exports.validationValidators = void 0;
var constValidator_1 = require("./constValidator");
var dependentRequiredValidator_1 = require("./dependentRequiredValidator");
var enumValidator_1 = require("./enumValidator");
var exclusiveMaximumValidator_1 = require("./exclusiveMaximumValidator");
var exclusiveMinimumValidator_1 = require("./exclusiveMinimumValidator");
var maxContainsValidator_1 = require("./maxContainsValidator");
var maxItemsValidator_1 = require("./maxItemsValidator");
var maxLengthValidator_1 = require("./maxLengthValidator");
var maxPropertiesValidator_1 = require("./maxPropertiesValidator");
var maximumValidator_1 = require("./maximumValidator");
var minContainsValidator_1 = require("./minContainsValidator");
var minItemsValidator_1 = require("./minItemsValidator");
var minLengthValidator_1 = require("./minLengthValidator");
var minPropertiesValidator_1 = require("./minPropertiesValidator");
var minimumValidator_1 = require("./minimumValidator");
var multipleOfValidator_1 = require("./multipleOfValidator");
var patternValidator_1 = require("./patternValidator");
var requiredValidator_1 = require("./requiredValidator");
var typeValidator_1 = require("./typeValidator");
var uniqueItemsValidator_1 = require("./uniqueItemsValidator");
exports.validationValidators = {
    type: typeValidator_1.typeValidator,
    "enum": enumValidator_1.enumValidator,
    "const": constValidator_1.constValidator,
    pattern: patternValidator_1.patternValidator,
    minLength: minLengthValidator_1.minLengthValidator,
    maxLength: maxLengthValidator_1.maxLengthValidator,
    exclusiveMaximum: exclusiveMaximumValidator_1.exclusiveMaximumValidator,
    multipleOf: multipleOfValidator_1.multipleOfValidator,
    exclusiveMinimum: exclusiveMinimumValidator_1.exclusiveMinimumValidator,
    maximum: maximumValidator_1.maximumValidator,
    minimum: minimumValidator_1.minimumValidator,
    dependentRequired: dependentRequiredValidator_1.dependentRequiredValidator,
    maxProperties: maxPropertiesValidator_1.maxPropertiesValidator,
    minProperties: minPropertiesValidator_1.minPropertiesValidator,
    required: requiredValidator_1.requiredValidator,
    maxItems: maxItemsValidator_1.maxItemsValidator,
    minItems: minItemsValidator_1.minItemsValidator,
    maxContains: maxContainsValidator_1.maxContainsValidator,
    minContains: minContainsValidator_1.minContainsValidator,
    uniqueItems: uniqueItemsValidator_1.uniqueItemsValidator
};
