// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/validation/maxItemsValidator.js
"use strict";
exports.__esModule = true;
exports.maxItemsValidator = void 0;
var isJSONArray_1 = require("../../../../util/isJSONArray");
function maxItemsValidator(schema, schemaPath, context) {
    if (!('maxItems' in schema)) {
        return null;
    }
    var maxItems = schema['maxItems'];
    var outputFormat = context.outputFormat;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONArray_1.isJSONArray)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        if (instance.length <= maxItems) {
            return { valid: true, schemaLocation: schemaLocation, schemaKeyword: 'maxItems', instanceLocation: instanceLocation };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'maxItems',
                    instanceLocation: instanceLocation,
                    message: maxItems === 1
                        ? "should have up to 1 item but has ".concat(instance.length, " instead")
                        : "should have up to ".concat(maxItems, " items but has ").concat(instance.length, " instead")
                };
            }
        }
    };
}
exports.maxItemsValidator = maxItemsValidator;
