// @criteria/json-schema-validation/dist/specification/draft-2020-12/vocabularies/validation/minItemsValidator.js
"use strict";
exports.__esModule = true;
exports.minItemsValidator = void 0;
var isJSONArray_1 = require("../../../../util/isJSONArray");
function minItemsValidator(schema, schemaPath, context) {
    if (!('minItems' in schema)) {
        return null;
    }
    var minItems = schema['minItems'];
    var outputFormat = context.outputFormat;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONArray_1.isJSONArray)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        if (instance.length >= minItems) {
            return { valid: true, schemaLocation: schemaLocation, schemaKeyword: 'minItems', instanceLocation: instanceLocation };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'minItems',
                    instanceLocation: instanceLocation,
                    message: minItems === 1
                        ? "should have at least 1 item but has ".concat(instance.length, " instead")
                        : "should have at least ".concat(minItems, " items but has ").concat(instance.length, " instead")
                };
            }
        }
    };
}
exports.minItemsValidator = minItemsValidator;
