// @hackbg/miscreant-esm/src/index.dist.cjs
"use strict";
var __createBinding = this && this.__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);
  if (!desc || (("get" in desc) ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function () {
        return m[k];
      }
    };
  }
  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});
var __exportStar = this && this.__exportStar || (function (m, exports) {
  for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
});
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebCryptoProvider = exports.SoftCryptoProvider = exports.PMAC = exports.CMAC = exports.StreamDecryptor = exports.StreamEncryptor = exports.SIV = exports.AEAD = void 0;
__exportStar(require("./interfaces.dist.cjs"), exports);
__exportStar(require("./exceptions.dist.cjs"), exports);
var aead_1 = require("./aead.dist.cjs");
Object.defineProperty(exports, "AEAD", {
  enumerable: true,
  get: function () {
    return aead_1.AEAD;
  }
});
var siv_1 = require("./siv.dist.cjs");
Object.defineProperty(exports, "SIV", {
  enumerable: true,
  get: function () {
    return siv_1.SIV;
  }
});
var stream_1 = require("./stream.dist.cjs");
Object.defineProperty(exports, "StreamEncryptor", {
  enumerable: true,
  get: function () {
    return stream_1.StreamEncryptor;
  }
});
Object.defineProperty(exports, "StreamDecryptor", {
  enumerable: true,
  get: function () {
    return stream_1.StreamDecryptor;
  }
});
var cmac_1 = require("./mac/cmac.dist.cjs");
Object.defineProperty(exports, "CMAC", {
  enumerable: true,
  get: function () {
    return cmac_1.CMAC;
  }
});
var pmac_1 = require("./mac/pmac.dist.cjs");
Object.defineProperty(exports, "PMAC", {
  enumerable: true,
  get: function () {
    return pmac_1.PMAC;
  }
});
var soft_1 = require("./providers/soft.dist.cjs");
Object.defineProperty(exports, "SoftCryptoProvider", {
  enumerable: true,
  get: function () {
    return soft_1.SoftCryptoProvider;
  }
});
var webcrypto_1 = require("./providers/webcrypto.dist.cjs");
Object.defineProperty(exports, "WebCryptoProvider", {
  enumerable: true,
  get: function () {
    return webcrypto_1.WebCryptoProvider;
  }
});
