// @criteria/json-schema/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
exports.__esModule = true;
exports.metaSchemaIDDraft2020_12 = exports.visitSubschemasDraft2020_12 = exports.dereferenceJSONSchemaDraft2020_12 = exports.metaSchemaIDDraft07 = exports.visitSubschemasDraft07 = exports.dereferenceJSONSchemaDraft07 = exports.metaSchemaIDDraft06 = exports.visitSubschemasDraft06 = exports.dereferenceJSONSchemaDraft06 = exports.metaSchemaIDDraft04 = exports.visitSubschemasDraft04 = exports.dereferenceJSONSchemaDraft04 = exports.dereferenceReferences = void 0;
var dereferenceReferences_1 = require("./dereferencing/dereferenceReferences");
__createBinding(exports, dereferenceReferences_1, "dereferenceReferences");
__exportStar(require("./dereferencing/mergeReference"), exports);
var draft_04_1 = require("./draft-04");
__createBinding(exports, draft_04_1, "dereferenceJSONSchema", "dereferenceJSONSchemaDraft04");
__createBinding(exports, draft_04_1, "visitSubschemas", "visitSubschemasDraft04");
__createBinding(exports, draft_04_1, "metaSchemaID", "metaSchemaIDDraft04");
var draft_06_1 = require("./draft-06");
__createBinding(exports, draft_06_1, "dereferenceJSONSchema", "dereferenceJSONSchemaDraft06");
__createBinding(exports, draft_06_1, "visitSubschemas", "visitSubschemasDraft06");
__createBinding(exports, draft_06_1, "metaSchemaID", "metaSchemaIDDraft06");
var draft_07_1 = require("./draft-07");
__createBinding(exports, draft_07_1, "dereferenceJSONSchema", "dereferenceJSONSchemaDraft07");
__createBinding(exports, draft_07_1, "visitSubschemas", "visitSubschemasDraft07");
__createBinding(exports, draft_07_1, "metaSchemaID", "metaSchemaIDDraft07");
var draft_2020_12_1 = require("./draft-2020-12");
__createBinding(exports, draft_2020_12_1, "dereferenceJSONSchema", "dereferenceJSONSchemaDraft2020_12");
__createBinding(exports, draft_2020_12_1, "visitSubschemas", "visitSubschemasDraft2020_12");
__createBinding(exports, draft_2020_12_1, "metaSchemaID", "metaSchemaIDDraft2020_12");
__exportStar(require("./retrievers"), exports);
__exportStar(require("./schema-index/DocumentIndex"), exports);
__exportStar(require("./schema-index/JSONReferenceContentIndex"), exports);
__exportStar(require("./schema-index/SchemaContentIndex"), exports);
__exportStar(require("./schema-index/SchemaIndex"), exports);
__exportStar(require("./schema-index/types"), exports);
