// @bufbuild/cel/dist/esm/gen/dev/cel/expr/operator_const.js
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// This file is generated by buf-yaml. DO NOT EDIT.
// generated from: proto/dev/cel/expr/operator.const.yaml
// TODO convert to enum
// Also see https://github.com/bufbuild/cel-es/blob/main/packages/cel/proto/dev/cel/expr/operator.const.yaml
export const ADD = "_+_";
export const ALL = "all";
export const CONDITIONAL = "_?_:_";
export const DIVIDE = "_/_";
export const EQUALS = "_==_";
export const EXISTS = "exists";
export const EXISTS_ONE = "exists_one";
export const FILTER = "filter";
export const GREATER = "_>_";
export const GREATER_EQUALS = "_>=_";
export const HAS = "has";
export const IN = "@in";
export const INDEX = "_[_]";
export const LESS = "_<_";
export const LESS_EQUALS = "_<=_";
export const LOGICAL_AND = "_&&_";
export const LOGICAL_NOT = "!_";
export const LOGICAL_OR = "_||_";
export const MAP = "map";
export const MODULO = "_%_";
export const MULTIPLY = "_*_";
export const NEGATE = "-_";
export const NOT_EQUALS = "_!=_";
export const NOT_STRICTLY_FALSE = "@not_strictly_false";
export const OLD_IN = "_in_";
export const OLD_NOT_STRICTLY_FALSE = "__not_strictly_false__";
export const OPT_INDEX = "_[?_]";
export const OPT_SELECT = "_?._";
export const SUBTRACT = "_-_";
