// @bufbuild/cel/dist/esm/gen/dev/cel/expr/overload_const.js
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// This file is generated by buf-yaml. DO NOT EDIT.
// generated from: proto/dev/cel/expr/overload.const.yaml
// TODO convert to enum
// Also see https://github.com/bufbuild/cel-es/blob/main/packages/cel/proto/dev/cel/expr/operator.const.yaml
export const ADD_BYTES = "add_bytes";
export const ADD_DOUBLE = "add_double";
export const ADD_DURATION_DURATION = "add_duration_duration";
export const ADD_DURATION_TIMESTAMP = "add_duration_timestamp";
export const ADD_INT64 = "add_int64";
export const ADD_LIST = "add_list";
export const ADD_STRING = "add_string";
export const ADD_TIMESTAMP_DURATION = "add_timestamp_duration";
export const ADD_UINT64 = "add_uint64";
export const BOOL_TO_BOOL = "bool_to_bool";
export const BOOL_TO_STRING = "bool_to_string";
export const BYTES_TO_BYTES = "bytes_to_bytes";
export const BYTES_TO_STRING = "bytes_to_string";
export const CONDITIONAL = "conditional";
export const CONTAINS = "contains";
export const CONTAINS_STRING = "contains_string";
export const DEPRECATED_IN = "in";
export const DIVIDE_DOUBLE = "divide_double";
export const DIVIDE_INT64 = "divide_int64";
export const DIVIDE_UINT64 = "divide_uint64";
export const DOUBLE_TO_DOUBLE = "double_to_double";
export const DOUBLE_TO_INT = "double_to_int64";
export const DOUBLE_TO_STRING = "double_to_string";
export const DOUBLE_TO_UINT = "double_to_uint64";
export const DURATION_TO_DURATION = "duration_to_duration";
export const DURATION_TO_HOURS = "duration_to_hours";
export const DURATION_TO_INT = "duration_to_int64";
export const DURATION_TO_MILLISECONDS = "duration_to_milliseconds";
export const DURATION_TO_MINUTES = "duration_to_minutes";
export const DURATION_TO_SECONDS = "duration_to_seconds";
export const DURATION_TO_STRING = "duration_to_string";
export const ENDS_WITH = "endsWith";
export const ENDS_WITH_STRING = "ends_with_string";
export const EQUALS = "equals";
export const GREATER_BOOL = "greater_bool";
export const GREATER_BYTES = "greater_bytes";
export const GREATER_DOUBLE = "greater_double";
export const GREATER_DOUBLE_INT64 = "greater_double_int64";
export const GREATER_DOUBLE_UINT64 = "greater_double_uint64";
export const GREATER_DURATION = "greater_duration";
export const GREATER_EQUALS_BOOL = "greater_equals_bool";
export const GREATER_EQUALS_BYTES = "greater_equals_bytes";
export const GREATER_EQUALS_DOUBLE = "greater_equals_double";
export const GREATER_EQUALS_DOUBLE_INT64 = "greater_equals_double_int64";
export const GREATER_EQUALS_DOUBLE_UINT64 = "greater_equals_double_uint64";
export const GREATER_EQUALS_DURATION = "greater_equals_duration";
export const GREATER_EQUALS_INT64 = "greater_equals_int64";
export const GREATER_EQUALS_INT64_DOUBLE = "greater_equals_int64_double";
export const GREATER_EQUALS_INT64_UINT64 = "greater_equals_int64_uint64";
export const GREATER_EQUALS_STRING = "greater_equals_string";
export const GREATER_EQUALS_TIMESTAMP = "greater_equals_timestamp";
export const GREATER_EQUALS_UINT64 = "greater_equals_uint64";
export const GREATER_EQUALS_UINT64_DOUBLE = "greater_equals_uint64_double";
export const GREATER_EQUALS_UINT64_INT64 = "greater_equals_uint64_int64";
export const GREATER_INT64 = "greater_int64";
export const GREATER_INT64_DOUBLE = "greater_int64_double";
export const GREATER_INT64_UINT64 = "greater_int64_uint64";
export const GREATER_STRING = "greater_string";
export const GREATER_TIMESTAMP = "greater_timestamp";
export const GREATER_UINT64 = "greater_uint64";
export const GREATER_UINT64_DOUBLE = "greater_uint64_double";
export const GREATER_UINT64_INT64 = "greater_uint64_int64";
export const HAS_NEXT = "@hasNext";
export const INDEX_LIST = "index_list";
export const INDEX_MAP = "index_map";
export const INDEX_MESSAGE = "index_message";
export const INT_TO_DOUBLE = "int64_to_double";
export const INT_TO_DURATION = "int64_to_duration";
export const INT_TO_INT = "int64_to_int64";
export const INT_TO_STRING = "int64_to_string";
export const INT_TO_TIMESTAMP = "int64_to_timestamp";
export const INT_TO_UINT = "int64_to_uint64";
export const IN_LIST = "in_list";
export const IN_MAP = "in_map";
export const IN_MESSAGE = "in_message";
export const ITERATOR = "@iterator";
export const LESS_BOOL = "less_bool";
export const LESS_BYTES = "less_bytes";
export const LESS_DOUBLE = "less_double";
export const LESS_DOUBLE_INT64 = "less_double_int64";
export const LESS_DOUBLE_UINT64 = "less_double_uint64";
export const LESS_DURATION = "less_duration";
export const LESS_EQUALS_BOOL = "less_equals_bool";
export const LESS_EQUALS_BYTES = "less_equals_bytes";
export const LESS_EQUALS_DOUBLE = "less_equals_double";
export const LESS_EQUALS_DOUBLE_INT64 = "less_equals_double_int64";
export const LESS_EQUALS_DOUBLE_UINT64 = "less_equals_double_uint64";
export const LESS_EQUALS_DURATION = "less_equals_duration";
export const LESS_EQUALS_INT64 = "less_equals_int64";
export const LESS_EQUALS_INT64_DOUBLE = "less_equals_int64_double";
export const LESS_EQUALS_INT64_UINT64 = "less_equals_int64_uint64";
export const LESS_EQUALS_STRING = "less_equals_string";
export const LESS_EQUALS_TIMESTAMP = "less_equals_timestamp";
export const LESS_EQUALS_UINT64 = "less_equals_uint64";
export const LESS_EQUALS_UINT64_DOUBLE = "less_equals_uint64_double";
export const LESS_EQUALS_UINT64_INT64 = "less_equals_uint64_int64";
export const LESS_INT64 = "less_int64";
export const LESS_INT64_DOUBLE = "less_int64_double";
export const LESS_INT64_UINT64 = "less_int64_uint64";
export const LESS_STRING = "less_string";
export const LESS_TIMESTAMP = "less_timestamp";
export const LESS_UINT64 = "less_uint64";
export const LESS_UINT64_DOUBLE = "less_uint64_double";
export const LESS_UINT64_INT64 = "less_uint64_int64";
export const LOGICAL_AND = "logical_and";
export const LOGICAL_NOT = "logical_not";
export const LOGICAL_OR = "logical_or";
export const MATCHES = "matches";
export const MATCHES_STRING = "matches_string";
export const MODULO_INT64 = "modulo_int64";
export const MODULO_UINT64 = "modulo_uint64";
export const MULTIPLY_DOUBLE = "multiply_double";
export const MULTIPLY_INT64 = "multiply_int64";
export const MULTIPLY_UINT64 = "multiply_uint64";
export const NEGATE_DOUBLE = "negate_double";
export const NEGATE_INT64 = "negate_int64";
export const NEXT = "@next";
export const NOT_EQUALS = "not_equals";
export const NOT_STRICTLY_FALSE = "not_strictly_false";
export const SIZE = "size";
export const SIZE_BYTES = "size_bytes";
export const SIZE_BYTES_INST = "bytes_size";
export const SIZE_LIST = "size_list";
export const SIZE_LIST_INST = "list_size";
export const SIZE_MAP = "size_map";
export const SIZE_MAP_INST = "map_size";
export const SIZE_STRING = "size_string";
export const SIZE_STRING_INST = "string_size";
export const STARTS_WITH = "startsWith";
export const STARTS_WITH_STRING = "starts_with_string";
export const STRING_TO_BOOL = "string_to_bool";
export const STRING_TO_BYTES = "string_to_bytes";
export const STRING_TO_DOUBLE = "string_to_double";
export const STRING_TO_DURATION = "string_to_duration";
export const STRING_TO_INT = "string_to_int64";
export const STRING_TO_STRING = "string_to_string";
export const STRING_TO_TIMESTAMP = "string_to_timestamp";
export const STRING_TO_UINT = "string_to_uint64";
export const SUBTRACT_DOUBLE = "subtract_double";
export const SUBTRACT_DURATION_DURATION = "subtract_duration_duration";
export const SUBTRACT_INT64 = "subtract_int64";
export const SUBTRACT_TIMESTAMP_DURATION = "subtract_timestamp_duration";
export const SUBTRACT_TIMESTAMP_TIMESTAMP = "subtract_timestamp_timestamp";
export const SUBTRACT_UINT64 = "subtract_uint64";
export const TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED = "timestamp_to_day_of_month_1_based";
export const TIMESTAMP_TO_DAY_OF_MONTH_ONE_BASED_WITH_TZ = "timestamp_to_day_of_month_1_based_with_tz";
export const TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED = "timestamp_to_day_of_month";
export const TIMESTAMP_TO_DAY_OF_MONTH_ZERO_BASED_WITH_TZ = "timestamp_to_day_of_month_with_tz";
export const TIMESTAMP_TO_DAY_OF_WEEK = "timestamp_to_day_of_week";
export const TIMESTAMP_TO_DAY_OF_WEEK_WITH_TZ = "timestamp_to_day_of_week_with_tz";
export const TIMESTAMP_TO_DAY_OF_YEAR = "timestamp_to_day_of_year";
export const TIMESTAMP_TO_DAY_OF_YEAR_WITH_TZ = "timestamp_to_day_of_year_with_tz";
export const TIMESTAMP_TO_HOURS = "timestamp_to_hours";
export const TIMESTAMP_TO_HOURS_WITH_TZ = "timestamp_to_hours_with_tz";
export const TIMESTAMP_TO_INT = "timestamp_to_int64";
export const TIMESTAMP_TO_MILLISECONDS = "timestamp_to_milliseconds";
export const TIMESTAMP_TO_MILLISECONDS_WITH_TZ = "timestamp_to_milliseconds_with_tz";
export const TIMESTAMP_TO_MINUTES = "timestamp_to_minutes";
export const TIMESTAMP_TO_MINUTES_WITH_TZ = "timestamp_to_minutes_with_tz";
export const TIMESTAMP_TO_MONTH = "timestamp_to_month";
export const TIMESTAMP_TO_MONTH_WITH_TZ = "timestamp_to_month_with_tz";
export const TIMESTAMP_TO_SECONDS = "timestamp_to_seconds";
export const TIMESTAMP_TO_SECONDS_WITH_TZ = "timestamp_to_seconds_tz";
export const TIMESTAMP_TO_STRING = "timestamp_to_string";
export const TIMESTAMP_TO_TIMESTAMP = "timestamp_to_timestamp";
export const TIMESTAMP_TO_YEAR = "timestamp_to_year";
export const TIMESTAMP_TO_YEAR_WITH_TZ = "timestamp_to_year_with_tz";
export const TIME_GET_DATE = "getDate";
export const TIME_GET_DAY_OF_MONTH = "getDayOfMonth";
export const TIME_GET_DAY_OF_WEEK = "getDayOfWeek";
export const TIME_GET_DAY_OF_YEAR = "getDayOfYear";
export const TIME_GET_FULL_YEAR = "getFullYear";
export const TIME_GET_HOURS = "getHours";
export const TIME_GET_MILLISECONDS = "getMilliseconds";
export const TIME_GET_MINUTES = "getMinutes";
export const TIME_GET_MONTH = "getMonth";
export const TIME_GET_SECONDS = "getSeconds";
export const TO_DYN = "to_dyn";
export const TYPE_CONVERT_BOOL = "bool";
export const TYPE_CONVERT_BYTES = "bytes";
export const TYPE_CONVERT_DOUBLE = "double";
export const TYPE_CONVERT_DURATION = "duration";
export const TYPE_CONVERT_DYN = "dyn";
export const TYPE_CONVERT_INT = "int";
export const TYPE_CONVERT_STRING = "string";
export const TYPE_CONVERT_TIMESTAMP = "timestamp";
export const TYPE_CONVERT_TYPE = "type";
export const TYPE_CONVERT_UINT = "uint";
export const UINT_TO_DOUBLE = "uint64_to_double";
export const UINT_TO_INT = "uint64_to_int64";
export const UINT_TO_STRING = "uint64_to_string";
export const UINT_TO_UINT = "uint64_to_uint64";
