// @criteria/json-schema-validation/dist/specification/draft-06/vocabularies/validation/exclusiveMaximumValidator.js
"use strict";
exports.__esModule = true;
exports.exclusiveMaximumValidator = void 0;
var format_1 = require("../../../../util/format");
var isJSONNumber_1 = require("../../../../util/isJSONNumber");
function exclusiveMaximumValidator(schema, schemaPath, context) {
    if (!('exclusiveMaximum' in schema)) {
        return null;
    }
    var exclusiveMaximum = schema['exclusiveMaximum'];
    var outputFormat = context.outputFormat;
    var schemaLocation = schemaPath.join('');
    return function (instance, instanceLocation, annotationResults) {
        if (!(0, isJSONNumber_1.isJSONNumber)(instance)) {
            return { valid: true, schemaLocation: schemaLocation, instanceLocation: instanceLocation };
        }
        if (instance < exclusiveMaximum) {
            return { valid: true, schemaLocation: schemaLocation, schemaKeyword: 'exclusiveMaximum', instanceLocation: instanceLocation };
        }
        else {
            if (outputFormat === 'flag') {
                return { valid: false };
            }
            else {
                return {
                    valid: false,
                    schemaLocation: schemaLocation,
                    schemaKeyword: 'exclusiveMaximum',
                    instanceLocation: instanceLocation,
                    message: "should be less than ".concat(exclusiveMaximum, " but is ").concat((0, format_1.format)(instance), " instead")
                };
            }
        }
    };
}
exports.exclusiveMaximumValidator = exclusiveMaximumValidator;
