// @confluentinc/schemaregistry/dist/google/type/dayofweek_pb.js
"use strict";
// Copyright 2021-2024 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DayOfWeekSchema = exports.DayOfWeek = exports.file_google_type_dayofweek = void 0;
const codegenv1_1 = require("@bufbuild/protobuf/codegenv1");
/**
 * Describes the file google/type/dayofweek.proto.
 */
exports.file_google_type_dayofweek = (0, codegenv1_1.fileDesc)("Chtnb29nbGUvdHlwZS9kYXlvZndlZWsucHJvdG8SC2dvb2dsZS50eXBlKoQBCglEYXlPZldlZWsSGwoXREFZX09GX1dFRUtfVU5TUEVDSUZJRUQQABIKCgZNT05EQVkQARILCgdUVUVTREFZEAISDQoJV0VETkVTREFZEAMSDAoIVEhVUlNEQVkQBBIKCgZGUklEQVkQBRIMCghTQVRVUkRBWRAGEgoKBlNVTkRBWRAHQmkKD2NvbS5nb29nbGUudHlwZUIORGF5T2ZXZWVrUHJvdG9QAVo+Z29vZ2xlLmdvbGFuZy5vcmcvZ2VucHJvdG8vZ29vZ2xlYXBpcy90eXBlL2RheW9md2VlaztkYXlvZndlZWuiAgNHVFBiBnByb3RvMw");
/**
 * Represents a day of the week.
 *
 * @generated from enum google.type.DayOfWeek
 */
var DayOfWeek;
(function (DayOfWeek) {
    /**
     * The day of the week is unspecified.
     *
     * @generated from enum value: DAY_OF_WEEK_UNSPECIFIED = 0;
     */
    DayOfWeek[DayOfWeek["DAY_OF_WEEK_UNSPECIFIED"] = 0] = "DAY_OF_WEEK_UNSPECIFIED";
    /**
     * Monday
     *
     * @generated from enum value: MONDAY = 1;
     */
    DayOfWeek[DayOfWeek["MONDAY"] = 1] = "MONDAY";
    /**
     * Tuesday
     *
     * @generated from enum value: TUESDAY = 2;
     */
    DayOfWeek[DayOfWeek["TUESDAY"] = 2] = "TUESDAY";
    /**
     * Wednesday
     *
     * @generated from enum value: WEDNESDAY = 3;
     */
    DayOfWeek[DayOfWeek["WEDNESDAY"] = 3] = "WEDNESDAY";
    /**
     * Thursday
     *
     * @generated from enum value: THURSDAY = 4;
     */
    DayOfWeek[DayOfWeek["THURSDAY"] = 4] = "THURSDAY";
    /**
     * Friday
     *
     * @generated from enum value: FRIDAY = 5;
     */
    DayOfWeek[DayOfWeek["FRIDAY"] = 5] = "FRIDAY";
    /**
     * Saturday
     *
     * @generated from enum value: SATURDAY = 6;
     */
    DayOfWeek[DayOfWeek["SATURDAY"] = 6] = "SATURDAY";
    /**
     * Sunday
     *
     * @generated from enum value: SUNDAY = 7;
     */
    DayOfWeek[DayOfWeek["SUNDAY"] = 7] = "SUNDAY";
})(DayOfWeek || (exports.DayOfWeek = DayOfWeek = {}));
/**
 * Describes the enum google.type.DayOfWeek.
 */
exports.DayOfWeekSchema = (0, codegenv1_1.enumDesc)(exports.file_google_type_dayofweek, 0);
