// @bufbuild/cel/dist/esm/activation.js
// Copyright 2024-2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { isCelError } from "./error.js";
import { toCel } from "./value.js";
export class ObjectActivation {
    constructor(data) {
        this.data = data;
    }
    resolve(name) {
        const value = this.data[name];
        if (value === undefined) {
            return undefined;
        }
        if (isCelError(value)) {
            return value;
        }
        return toCel(value);
    }
}
export class VarActivation {
    constructor(name, value, parent) {
        this.name = name;
        this.value = value;
        this.parent = parent;
    }
    resolve(name) {
        if (name === this.name) {
            return this.value;
        }
        return this.parent.resolve(name);
    }
}
export const EMPTY_ACTIVATION = {
    resolve() {
        return undefined;
    },
};
